### ISSUE TITLE
API REST and Reserved Order Id

### ISSUE THREAD

**Issue by davidlhoumaud** [magento/magento2#19101]
### Preconditions (*)
1. Magento CE 2.2.6 without sample data is installed
2.  PHP 7.0.1

### Steps to reproduce (*)
1. PUT REQUEST
2. http://magento.host/rest/V1/carts/mine
3. in the cart you must add a reserved_order_id (for example: 000000653)
 
### Expected result (*)
Update the cart with a reserved order number with the left pad 0 eg: 000000651

### Actual result (*)
Can not update cart with a reserved order number

### My temporary solution

in the file [CartInterface.php](https://github.com/magento/magento2/blob/53a896645c8a1a55fe7842ce621a1f79176b6d38/app/code/Magento/Quote/Api/Data/CartInterface.php)
There is a typing error Doctrine
The type should be a string of characters and not integer
```
/**
     * Returns the reserved order ID for the cart.
     *
     * @return int|null Reserved order ID. Otherwise, null.
     */
    public function getReservedOrderId();

    /**
     * Sets the reserved order ID for the cart.
     *
     * @param int $reservedOrderId
     * @return $this
     */
    public function setReservedOrderId($reservedOrderId);
```
We must replace the comments Doctrine by here

```
/**
     * Returns the reserved order ID for the cart.
     *
     * @return string|null Reserved order ID. Otherwise, null.
     */
    public function getReservedOrderId();

    /**
     * Sets the reserved order ID for the cart.
     *
     * @param string $reservedOrderId
     * @return $this
     */
    public function setReservedOrderId($reservedOrderId);
```
I make the patch with the following SED command
`cat vendor/magento/module-quote/Api/Data/CartInterface.php" | sed -e "s/int\|null Reserved order ID/string\|null Reserved order ID/g" -e "s/int \$reservedOrderId/string \$reservedOrderId/g"`

**Comment by magento-engcom-team** [comment]
Hi @davidlhoumaud. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@davidlhoumaud do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by davidlhoumaud** [comment]
@magento-engcom-team give me 2.2-develop instance

**Comment by magento-engcom-team** [comment]
Hi @davidlhoumaud. Thank you for your request. I'm working on Magento 2.2-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @davidlhoumaud, here is your [Magento instance](http://34.228.235.121/i-19101-2-2-develop/).
Admin access: [http://34.228.235.121/i-19101-2-2-develop//admin](http://34.228.235.121/i-19101-2-2-develop//admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by davidlhoumaud** [comment]
Confirmed this on vanilla instance.

**Comment by magento-engcom-team** [comment]
Hi @rogyar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@rogyar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96237`, `MAGETWO-96238` were created

**Comment by magento-engcom-team** [comment]
Hi @vasilii-b. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [x] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [x] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by magento-engcom-team** [comment]
Hi @davidlhoumaud. Thank you for your report. 
The issue has been fixed in magento/magento2#19130 by @vasilii-b in 2.3-develop branch
Related commit(s):
 - [fe7b9ac55097b64a023e607acacb1857f1926367](https://github.com/magento/magento2/commit/fe7b9ac55097b64a023e607acacb1857f1926367)

The fix will be available with the upcoming 2.3.1 release.

**Comment by magento-engcom-team** [comment]
Hi @davidlhoumaud. Thank you for your report. 
The issue has been fixed in magento/magento2#20208 by @saphaljha in 2.2-develop branch
Related commit(s):
 - [e3ec0f68f55f9268c03c4c25152b4cc2c8be737c](https://github.com/magento/magento2/commit/e3ec0f68f55f9268c03c4c25152b4cc2c8be737c)

The fix will be available with the upcoming 2.2.8 release.
